### PR TITLE
[C] Add convenience functions

### LINKED ISSUE #69: [C] Move the ArrowArrayStream helpers from C++ to C?
The EmptyArrayStream/VectorArrayStream may also be useful in C, would it be useful to move them to be implemented in C (with C++ API wrappers)? I've found myself writing similar utilities in C for the ADBC SQLite driver

### LINKED ISSUE #70: [C] Add convenience functions
...based on writing some nanoarrow code in the R package, https://github.com/geoarrow/geoarrow-cpp/ , and https://github.com/paleolimbot/minigpkg/

A few that may be nice are:

- `ArrowBufferAppendStringView()` (i.e., `ArrowBufferAppendStringView(buffer, ArrowCharView("some string");`)
- `ArrowBufferAppendBufferView()` (i.e., `ArrowBufferAppendBufferView(buffer, some_array_view->buffer_views[1])`)
- `ArrowArrayMove(struct ArrowArray* src, struct ArrowArray* dst)`, `ArrowSchemaMove(struct ArrowSchema* src, struct ArrowSchema* dst)`, and `ArrowArrayStreamMove(struct ArrowArrayStream* src, struct ArrowArrayStream* dst)` (more readable than `memcpy(dst, src, sizeof(struct ArrowSomething); src->release = NULL;`)

Another thing that came up is a pattern for unsafe appends. The buffer appenders in the `ArrowBufferAppendInt32()` family all check and possibly reallocate; the array appenders in the `ArrowArrayAppendInt()` family do a `switch()` on type and range checks to help prevent implicit overflow. Adding more convenience around this like maybe `ArrowBufferAppendInt32Unsafe()` or `ArrowArrayInt32AppendInt32Unsafe()` is probably very verbose and isn't that hard to work around. I could be convinced either way on whether they should exist or not.

### PR DESCRIPTION
**Pull request by paleolimbot**
Fixes #69; fixes #70.

- [x] ArrowXXXMove()
- [x] ArrowBufferAppend(String|Buffer)View()
- [x] Basic array stream implementation for a finite number of previously-allocated arrays

Still needs:

- [x] Rework nanoarrow.hpp ArrayStream helpers to use the C version
- [x] More tests for the array stream (incompletely drained, empty, invalid arrays)

### CODE REVIEW

File: src/nanoarrow/nanoarrow.h [hunk 1]
```diff
@@ -851,6 +877,43 @@ static inline struct ArrowBufferView ArrowArrayViewGetBytesUnsafe(
...
+/// be greater than zero and less than the value of n_arrays passed in
+/// ArrowBasicArrayStreamInit(). Callers are not required to fill all
+/// n_arrays members (i.e., n_arrays is a maximum bound).
+void ArrowBasicArrayStreamSetArray(struct ArrowArrayStream* array_stream, int64_t i,
+                                   struct ArrowArray* array);
+
+/// \brief Validate the contents of this ArrowArrayStream
+///
+/// array_stream must have been initialized with ArrowBasicArrayStreamInit().
+/// This function uses ArrowArrayStreamInitFromSchema() and ArrowArrayStreamSetAray()
```
Comment (lidavidm) [inline src/nanoarrow/nanoarrow.h:910]:
```suggestion
/// This function uses ArrowArrayStreamInitFromSchema() and ArrowArrayStreamSetArray()
```

**Review by lidavidm — APPROVED** [review]
